File: services/learning/src/common/decimal-str.ts
/**
 * Tipo lógico `DecimalString` de los contratos (research D-10): la
 * representación canónica con la que TODO monto, tasa, peso o calificación cruza
 * una frontera —gRPC, REST o JSON—.
 *
 * El Principio VIII (NON-NEGOTIABLE) prohíbe `number` para estos valores: un
 * `number` de JavaScript es un IEEE-754 de doble precisión, así que `0.1 + 0.2`
 * no da `0.3` y una calificación o un monto en COP puede perder centavos. El
 * transporte es por tanto `string`, y este módulo es el único lugar del servicio
 * donde esa `string` se convierte a `Decimal` y vuelta.
 *
 * El formato canónico es EXACTAMENTE `^-?\d+(\.\d+)?$`:
 *
 * | entrada        | resultado                            |
 * |----------------|--------------------------------------|
 * | `"1500000.00"` | válido                               |
 * | `"-0.5"`       | válido                               |
 * | `"1.5e3"`      | RECHAZADO — notación científica      |
 * | `"1,500.00"`   | RECHAZADO — separador de miles       |
 * | `"+1.5"`       | RECHAZADO — signo positivo explícito |
 * | `".5"` / `"5."`| RECHAZADO — falta un lado del punto  |
 * | `" 1.5"`       | RECHAZADO — espacios                 |
 *
 * Rechazar en lugar de normalizar es deliberado: si dos servicios discrepan en
 * la escala o el formato de un valor, queremos un error en la frontera y no un
 * dato silenciosamente distinto en la base de datos.
 */
import Decimal from 'decimal.js';

/**
 * Cantidad de dígitos (precisión o escala de una columna `NUMERIC`).
 *
 * Existe para poder tipar los contadores de dígitos sin usar `number` en cada
 * firma: el lint de este archivo prohíbe el tipo `number` (Principio VIII), y con
 * razón, porque aquí `number` casi siempre sería un valor monetario mal tipado.
 * Un contador de dígitos NO es un valor financiero —es un entero pequeño que
 * describe una columna— así que la excepción se concentra en esta única línea,
 * donde queda documentada y auditable, en lugar de repartirse por el archivo.
 */
// eslint-disable-next-line @typescript-eslint/no-restricted-types -- ver el comentario de arriba: contador de dígitos, no un valor monetario
export type DigitCount = number;

/** Límites de las columnas `NUMERIC` de data-model.md §Convenciones. */
export const LIMITS = {
  /** Montos (COP): `NUMERIC(19,2)`. */
  money: { precision: 19 as DigitCount, scale: 2 as DigitCount },
  /** Tasas y porcentajes: `NUMERIC(9,6)`. */
  rate: { precision: 9 as DigitCount, scale: 6 as DigitCount },
  /** Calificaciones, pesos y umbrales: `NUMERIC(6,2)`. */
  score: { precision: 6 as DigitCount, scale: 2 as DigitCount },
} as const;

/** Motivo por el que una `DecimalString` fue rechazada. */
export type DecimalStrErrorCode =
  /** Cadena vacía. */
  | 'empty'
  /** No respeta `^-?\d+(\.\d+)?$`. */
  | 'syntax'
  /** Más decimales significativos de los que admite la columna destino. */
  | 'scale'
  /** La parte entera no cabe en la columna destino. */
  | 'range';

/**
 * Error de interpretación o serialización de una `DecimalString`.
 *
 * Lleva un `code` discriminante para que la capa gRPC distinga un problema de
 * formato (dato mal construido por el emisor) de uno de rango (dato bien formado
 * que no cabe en la columna): ameritan respuestas distintas.
 */
export class DecimalStrError extends Error {
  public readonly code: DecimalStrErrorCode;

  public constructor(code: DecimalStrErrorCode, message: string) {
    super(message);
    this.name = 'DecimalStrError';
    this.code = code;
  }
}

/**
 * Única sintaxis aceptada.
 *
 * Se comprueba ANTES de entregar la cadena a decimal.js: su constructor acepta
 * de buen grado notación científica (`"1e5"`), signo positivo, `"Infinity"` y
 * `"NaN"`, que aquí no son canónicos.
 */
const CANONICAL = /^-?\d+(\.\d+)?$/;

/**
 * Convierte una cadena decimal canónica en `Decimal`.
 *
 * No impone límite de precisión: para validar contra una columna concreta usar
 * {@link parseMoney}, {@link parseRate}, {@link parseScore} o
 * {@link parseNumeric}.
 *
 * @throws {DecimalStrError} con code `empty` o `syntax`.
 */
export function parse(s: string): Decimal {
  if (s === '') {
    throw new DecimalStrError('empty', 'decimal-str: cadena vacía');
  }
  if (!CANONICAL.test(s)) {
    throw new DecimalStrError(
      'syntax',
      `decimal-str: formato no canónico: ${JSON.stringify(s)}`,
    );
  }
  return new Decimal(s);
}

/**
 * Como {@link parse}, y además exige que el valor quepa en una columna
 * `NUMERIC(precision, scale)` de PostgreSQL.
 *
 * La escala se mide sobre los decimales SIGNIFICATIVOS: `"1.500"` cuenta como
 * escala 1, no 3, porque los ceros a la derecha no aportan precisión y
 * rechazarlos solo castigaría a un emisor que rellena a un ancho fijo.
 *
 * @throws {DecimalStrError} con code `empty`, `syntax`, `scale` o `range`.
 */
export function parseNumeric(
  s: string,
  precision: DigitCount,
  scale: DigitCount,
): Decimal {
  const d = parse(s);

  const got = significantScale(s);
  if (got > scale) {
    throw new DecimalStrError(
      'scale',
      `decimal-str: escala excedida: ${JSON.stringify(s)} tiene ${got} decimales, el máximo es ${scale}`,
    );
  }

  // Cota exacta de PostgreSQL: |valor| < 10^(precision-scale).
  const maxAbs = new Decimal(10).pow(precision - scale);
  if (d.abs().gte(maxAbs)) {
    throw new DecimalStrError(
      'range',
      `decimal-str: fuera de rango: ${JSON.stringify(s)} excede NUMERIC(${precision},${scale})`,
    );
  }

  return d;
}

/**
 * Valida un monto contra `NUMERIC(19,2)`.
 *
 * @throws {DecimalStrError} ver {@link parseNumeric}.
 */
export function parseMoney(s: string): Decimal {
  return parseNumeric(s, LIMITS.money.precision, LIMITS.money.scale);
}

/**
 * Valida una tasa o porcentaje contra `NUMERIC(9,6)`.
 *
 * @throws {DecimalStrError} ver {@link parseNumeric}.
 */
export function parseRate(s: string): Decimal {
  return parseNumeric(s, LIMITS.rate.precision, LIMITS.rate.scale);
}

/**
 * Valida una calificación, peso o umbral contra `NUMERIC(6,2)`.
 *
 * @throws {DecimalStrError} ver {@link parseNumeric}.
 */
export function parseScore(s: string): Decimal {
  return parseNumeric(s, LIMITS.score.precision, LIMITS.score.scale);
}

/**
 * Serializa un `Decimal` a la forma canónica.
 *
 * Usa `toFixed()` sin argumentos y no `toString()`: decimal.js pasa a notación
 * exponencial en `toString()` a partir de cierto tamaño (`"1e+21"`), lo que
 * rompería a cualquier consumidor del contrato. `toFixed()` nunca lo hace.
 */
export function format(d: Decimal): string {
  const s = d.toFixed();
  // decimal.js conserva el signo del cero; "-0" no es canónico.
  return s === '-0' ? '0' : s;
}

/**
 * Serializa con exactamente `scale` decimales, rellenando con ceros.
 *
 * @throws {DecimalStrError} con code `scale` si el valor tiene MÁS decimales
 * significativos que `scale`: redondear en la capa de serialización esconde una
 * pérdida de precisión que el llamador no pidió. Para redondear hay que hacerlo
 * explícito con {@link roundHalfEven}.
 */
export function formatFixed(d: Decimal, scale: DigitCount): string {
  const got = significantScale(format(d));
  if (got > scale) {
    throw new DecimalStrError(
      'scale',
      `decimal-str: el valor tiene ${got} decimales, se pidieron ${scale}; usar roundHalfEven`,
    );
  }
  return d.toFixed(scale);
}

/**
 * Redondea a `scale` decimales con redondeo bancario (half-even), el único modo
 * permitido para conversiones y cálculos monetarios (research D-14).
 *
 * Es explícito a propósito: ninguna otra función de este módulo redondea.
 */
export function roundHalfEven(d: Decimal, scale: DigitCount): Decimal {
  return d.toDecimalPlaces(scale, Decimal.ROUND_HALF_EVEN);
}

/**
 * Cuenta los decimales de una representación canónica ignorando los ceros
 * finales.
 *
 * Se calcula sobre la cadena y no sobre el `Decimal` porque lo que interesa es
 * la precisión que el emisor declaró, antes de cualquier normalización.
 */
function significantScale(s: string): DigitCount {
  const dot = s.indexOf('.');
  if (dot < 0) {
    return 0;
  }
  return s.slice(dot + 1).replace(/0+$/, '').length;
}

File: services/simulator/tests/decimal_str.rs
//! Pruebas del tipo lógico `DecimalString` (T023).
//!
//! Cubren montos extremos, escala máxima, notación científica rechazada y
//! overflow, más la propiedad que motiva el Principio VIII: la aritmética debe
//! ser exacta justo donde `f64` falla.

use fintcart_simulator::domain::decimal_str::{
    format, format_fixed, parse, parse_money, parse_rate, parse_score, round_half_even,
    DecimalStrError,
};
use rust_decimal::Decimal;

/// Atajo para las pruebas: parsea o hace pánico con contexto.
fn d(s: &str) -> Decimal {
    parse(s).unwrap_or_else(|e| panic!("parse({s:?}) falló: {e}"))
}

#[test]
fn parse_acepta_la_forma_canonica() {
    for s in [
        "0",
        "-0",
        "5",
        "123",
        "0.5",
        "-0.5",
        "1500000.00",
        "-1500000.00",
        "0.000000000000000001",
    ] {
        assert!(parse(s).is_ok(), "parse({s:?}) debería aceptarse");
    }
}

#[test]
fn parse_rechaza_lo_no_canonico() {
    // Sintaxis: cada caso es una forma en que un emisor podría equivocarse.
    for s in [
        "1.5e3",    // notación científica
        "1.5E3",    // idem, mayúscula
        "1e5",      // idem, sin punto
        "1.5e-3",   // idem, exponente negativo
        "1,500.00", // separador de miles
        "1.500,00", // formato europeo
        "+1.5",     // signo positivo explícito
        ".5",       // sin parte entera
        "5.",       // sin parte decimal
        " 1.5",     // espacio inicial
        "1.5 ",     // espacio final
        "1 500",    // espacio interno
        "--1",      // doble signo
        "1.2.3",    // doble punto
        "0x10",     // hexadecimal
        "abc",      // texto
        "Infinity", // infinito
        "NaN",      // NaN
        "-",        // solo signo
        "1_000",    // guion bajo
        "$1500",    // símbolo de moneda
        "5%",       // porcentaje
        "1.5-",     // signo al final
    ] {
        match parse(s) {
            Err(DecimalStrError::Syntax(_)) => {}
            otro => panic!("parse({s:?}) debería dar Syntax, dio {otro:?}"),
        }
    }

    assert_eq!(parse(""), Err(DecimalStrError::Empty));
}

#[test]
fn parse_rechaza_lo_que_no_cabe_en_un_decimal() {
    // `rust_decimal` tiene mantisa de 96 bits: una parte entera de 40 dígitos no
    // es representable. Debe dar error, NO redondear en silencio ni entrar en
    // pánico. (En Go el equivalente usa big.Int y sí lo admite; la diferencia es
    // irrelevante porque toda columna del modelo cabe de sobra.)
    let enorme = "9".repeat(40);
    match parse(&enorme) {
        Err(DecimalStrError::Unrepresentable(_)) => {}
        otro => panic!("parse(40 nueves) debería dar Unrepresentable, dio {otro:?}"),
    }

    // Escala mayor que la máxima de Decimal (28) tampoco puede redondearse.
    let escala_absurda = format!("0.{}", "1".repeat(40));
    match parse(&escala_absurda) {
        Err(DecimalStrError::Unrepresentable(_)) => {}
        otro => panic!("parse(40 decimales) debería dar Unrepresentable, dio {otro:?}"),
    }
}

#[test]
fn parse_money_respeta_numeric_19_2() {
    // Aceptados: 17 dígitos enteros y 2 decimales es el máximo exacto.
    for s in [
        "0",
        "1500000.00",
        "99999999999999999.99",
        "-99999999999999999.99",
        "1.500", // los ceros de relleno no cuentan como escala
    ] {
        assert!(
            parse_money(s).is_ok(),
            "parse_money({s:?}) debería aceptarse"
        );
    }

    // Un dígito entero más ya no cabe.
    for s in ["100000000000000000.00", "-100000000000000000.00"] {
        match parse_money(s) {
            Err(DecimalStrError::Range { .. }) => {}
            otro => panic!("parse_money({s:?}) debería dar Range, dio {otro:?}"),
        }
    }

    // Tres decimales significativos exceden la escala de la columna.
    for s in ["1.001", "0.000000000000000001"] {
        match parse_money(s) {
            Err(DecimalStrError::Scale { .. }) => {}
            otro => panic!("parse_money({s:?}) debería dar Scale, dio {otro:?}"),
        }
    }
}

#[test]
fn parse_rate_y_parse_score_respetan_sus_columnas() {
    // Tasas: NUMERIC(9,6) → 3 enteros, 6 decimales.
    assert!(parse_rate("999.999999").is_ok());
    assert!(matches!(
        parse_rate("1000.0"),
        Err(DecimalStrError::Range { .. })
    ));
    assert!(matches!(
        parse_rate("0.0000001"),
        Err(DecimalStrError::Scale { .. })
    ));

    // Calificaciones: NUMERIC(6,2) → 4 enteros, 2 decimales.
    assert!(parse_score("100.00").is_ok());
    assert!(matches!(
        parse_score("10000.00"),
        Err(DecimalStrError::Range { .. })
    ));
}

#[test]
fn format_es_canonico_y_estable() {
    for (entrada, esperado) in [
        ("0", "0"),
        ("-0", "0"), // el cero negativo no es canónico en la salida
        ("5", "5"),
        ("1500000.00", "1500000"),
        ("1.500", "1.5"),
        ("-0.50", "-0.5"),
        ("100", "100"), // no se tocan los ceros de la parte entera
        ("0.010", "0.01"),
    ] {
        let salida = format(d(entrada));
        assert_eq!(salida, esperado, "format(parse({entrada:?}))");

        // Idempotencia: la salida canónica vuelve a parsearse igual.
        assert_eq!(format(d(&salida)), salida, "format no es idempotente");
    }
}

#[test]
fn format_nunca_usa_notacion_cientifica() {
    // Una serialización ingenua emitiría "1E+18" y rompería a cualquier
    // consumidor del contrato.
    for s in [
        "1000000000000000000",
        "0.000000000000000001",
        "-1000000000000000000",
    ] {
        let salida = format(d(s));
        assert!(
            !salida.contains('e') && !salida.contains('E'),
            "format({s:?}) = {salida:?} contiene notación científica"
        );
        assert!(
            parse(&salida).is_ok(),
            "format({s:?}) = {salida:?} no es canónico"
        );
    }
}

#[test]
fn format_fixed_rellena_pero_no_redondea_en_silencio() {
    assert_eq!(format_fixed(d("1.5"), 2).unwrap(), "1.50");
    assert_eq!(format_fixed(d("0"), 2).unwrap(), "0.00");

    // Pedir menos decimales de los que el valor tiene debe fallar, no redondear:
    // la pérdida silenciosa de precisión es lo que el Principio VIII evita.
    assert!(matches!(
        format_fixed(d("1.005"), 2),
        Err(DecimalStrError::Scale { .. })
    ));
}

#[test]
fn round_half_even_es_bancario() {
    for (entrada, escala, esperado) in [
        // El empate se resuelve hacia el dígito par, no siempre hacia arriba.
        ("0.125", 2, "0.12"),
        ("0.135", 2, "0.14"),
        ("2.5", 0, "2"),
        ("3.5", 0, "4"),
        ("-2.5", 0, "-2"),
        ("-0.125", 2, "-0.12"),
        // Sin empate se redondea normalmente.
        ("0.126", 2, "0.13"),
        ("0.124", 2, "0.12"),
    ] {
        let salida = format(round_half_even(d(entrada), escala));
        assert_eq!(salida, esperado, "round_half_even({entrada}, {escala})");
    }
}

/// La prueba del Principio VIII (SC-004): los casos clásicos en los que `f64`
/// da un resultado incorrecto deben ser exactos por esta vía.
#[test]
fn aritmetica_exacta_donde_el_float_falla() {
    // 0.1 + 0.2 != 0.3 en binario.
    assert_eq!(format(d("0.1") + d("0.2")), "0.3");

    // Un monto grande en COP más un centavo: con f64 el centavo se pierde por
    // falta de dígitos significativos.
    assert_eq!(
        format(d("99999999999999.99") + d("0.01")),
        "100000000000000"
    );

    // Diez veces 0.1 debe ser exactamente 1.
    let mut acc = Decimal::ZERO;
    for _ in 0..10 {
        acc += d("0.1");
    }
    assert_eq!(format(acc), "1");
}
